Allow to couple/uncouple CE pin

jjtmp opened the pull request.

jjtmp committed:
couple/uncouple (Nokia_LCD.h)

Part I/II

jjtmp committed:
couple/uncouple (Nokia_LCD.cpp)

Part II/II

To use couple function, just use "couple()" in the lcd initilization code. Then clock is always enabled which makes drawing a bit faster on regular pins and helps reducing interference when using two or more displays on the same data pins.

platisd commented on src/Nokia_LCD.h:
```diff
@@ -231,6 +231,14 @@ class Nokia_LCD {
+     * for accessing two or more displays on the same pins
+     * (but different CE pins)
+     */
+    void couple(bool yes = true);
+    void uncouple();
```
Is `couple(false)` the same as `uncouple`?
If so, then let's keep only one.

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -291,7 +307,8 @@ bool Nokia_LCD::send(const unsigned char lcd_byte, const bool is_data,
 
     // Send the byte
-    digitalWrite(kCe_pin, LOW);
+    if (!coupled)
+        digitalWrite(kCe_pin, LOW);
```
Consistent styling:

```suggestion
    if (!coupled) {
        digitalWrite(kCe_pin, LOW);
    }
```

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -112,6 +112,22 @@ void Nokia_LCD::begin() {
 }
 
+void Nokia_LCD::couple(bool yes) {
+    if (yes && !coupled) {
+        digitalWrite(kCe_pin, LOW);
+    }
+    else if (!yes && coupled) {
+        digitalWrite(kCe_pin, HIGH);
+    }
+    coupled = yes;
```
Hmm, couldn't this be changed to something simpler, i.e.:

```suggestion
    digitalWrite(kCe_pin, yes ? LOW : HIGH);
    coupled = yes;
```

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -301,7 +318,8 @@ bool Nokia_LCD::send(const unsigned char lcd_byte, const bool is_data,
     }
 
-    digitalWrite(kCe_pin, HIGH);
+    if (!coupled)
+        digitalWrite(kCe_pin, HIGH);
```
Consistent styling:
```suggestion
    if (!coupled) {
        digitalWrite(kCe_pin, HIGH);
    }
```

platisd reviewed (changes requested):
Hi Jonas, thanks for your contribution.

Can you explain _how_ would the new functions would be used in practice with a bare minimum example?

platisd commented on src/Nokia_LCD.h:
```diff
@@ -231,6 +231,14 @@ class Nokia_LCD {
      */
     void setDefaultFont();
 
+    /**
+     * couple()/uncouple() is useful for improving speed and 
+     * for accessing two or more displays on the same pins
+     * (but different CE pins)
+     */
```
The claims are that using these two new functions:
1. Improve speed
2. Allow accessing more than 1 display on the same pins

I can understand the 2nd point, but can you elaborate more on the 1st?

jjtmp replied:
Two answer all your remarks and maybe more:

1. Why this proposal

I was experimenting with two displays. But they constantly resisted to work as expected. I suspect electrical interference between the CE lines. Another advantage is that the contrast setting menu can be shown on all displays the same time so if your application misconfigured one LCD but shows the contrast setting menu on the other LCD too you still can read what's going on.


2. How-to use

The key of couple() is to stop the send() method from modifying the CE pin. Obviously you have to manage it by yourself then.

The bonus of couple() is that you can switch from one display to another without changing all these lcd.print() invocations in main and subroutines. I use this - but to be frank it would be more clear to wrap a digitalWrite such as activate_lcd1() {digitalWrite(CEofLCD1, LOW);} - I probably will switch to this soon for sake of clarity.


3. Examples

See
a) Full http://z-info.jbechtel.de/srv/23kopierer_savev179web.ino
b) Reduced http://z-info.jbechtel.de/srv/23kopierer_web_short.ino
c) Mini http://z-info.jbechtel.de/srv/23kopierer_web_veryshort.ino

The a) example is an actually working one. But it contains lots of code which you don't need for LCD (and also lots of dead code).
The b) example is untested. I just took the a) example and removed everything which is not contrast setup menu/counter printer.
The c) example is pretty minimal, just has the initialization phase of b) example. Also untested.

Please don't mind the "lcd.uncouple()" in all a)b)c) examples initialization. It is wrong and most probably should be just removed.
Also please don't mind that it is not neccessary to instantiate lcdL and lcdR objects. lcd object is just enough. All lcdL and lcdR tasks could be surrounded by activate_lcd1() like wrappers (see above 2.).


4. Improve speed

It was just my feeling that on Arduino the PIN access is pretty slow, compared to directly coding for ATMega. Web search engine results support this. (digitalWrite() seems to switch off PWM first, etc.). So omitting digitalWrite(), even just the surrounding one for CE, helps a bit. Of course one could be pretty much faster when not-at-all using digitalWrite() but reducint calls is not so bad, either.


I have incorporated most of your suggestions and expanded the documentation comment in Nokia_LCD.h. Is it better than before?

jjtmp committed:
Change couple() description and arguments (Nokia_LCD.h)

As suggested by pull request review

jjtmp committed:
Consistent code styling; reduce couple() arguments (Nokia_LCD.cpp)

As suggested by pull request review

jjtmp committed:
comment

platisd commented on src/Nokia_LCD.h:
```diff
@@ -231,6 +231,27 @@ class Nokia_LCD {
+     *            lcd.setInverted(true); lcd.print(number); lcd.setInverted(false);
+     *            lcd.print(" "); print_fwVersion(); lcd.println();
+     *     ac) reducing electrical interference in case it happens that one CE line
+     *         overspeaks to the other CE line
```
Sounds a bit farfetched but I haven't tried it so I guess you have a point.
However, it feels like not only the electrical engineering aspects are outside the scope of this library, but -more importantly- this isn't some information we should include in the API documentation.

platisd commented on src/Nokia_LCD.h:
```diff
@@ -231,6 +231,27 @@ class Nokia_LCD {
+     *         overspeaks to the other CE line
+     *     ad) manage CE pin by different code
+     *            lcd.couple(); // once and forever, now someone else has to manage
+     *                          // CE pin
```
I think I understand this one, but does it need to be in the API documentation?

platisd commented on src/Nokia_LCD.h:
```diff
@@ -231,6 +231,27 @@ class Nokia_LCD {
+     *     ad) manage CE pin by different code
+     *            lcd.couple(); // once and forever, now someone else has to manage
+     *                          // CE pin
+     * b) improving speed by ~10% (16 instead of 18 digitalWrite() per data/command byte)
```
While there are 10% less instructions that doesn't necessarily mean it's 10% faster, especially since branches have been introduced to the code.
Regardless, any performance claim should be backed up by data and I'd argue that even if there is a 10% performance boost, it isn't worth mentioning in the API. :)

platisd commented on src/Nokia_LCD.h:
```diff
@@ -231,6 +231,27 @@ class Nokia_LCD {
+     *            lcd.couple(); // once and forever, now someone else has to manage
+     *                          // CE pin
+     * b) improving speed by ~10% (16 instead of 18 digitalWrite() per data/command byte)
+     */
```
Based on the comments above, I suggest this instead:

```suggestion
    /**
     * @brief Allow the specific display instance to be controlled by another
     * instance, which has been initialized with the same pins except the CE one. 
     * Enables having multiple displays on the same data, clock and reset pins.
     * @example Allow `lcd1` and `lcd2 to both display the same as `lcd3`
     * @example `lcd1.couple(); lcd2.couple(); lcd3.print("Hi there");`
     */
```

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -112,6 +112,19 @@ void Nokia_LCD::begin() {
+    if (!coupled) {
+        digitalWrite(kCe_pin, LOW);
+        coupled = true;
+    }
```
Any particular reason why the `if` statement is _not_ skipped?

If the claim is performance, then it should be backed up by measurements to determine whether the increase in complexity justifies it.

```suggestion
    digitalWrite(kCe_pin, LOW);
    mCoupled = true;
```

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -112,6 +112,19 @@ void Nokia_LCD::begin() {
+    if (coupled) {
+        digitalWrite(kCe_pin, HIGH);
+        coupled = false;
+    }
```
Same as above:
```suggestion
    digitalWrite(kCe_pin, HIGH);
    mCoupled = false;
```

platisd commented on src/Nokia_LCD.h:
```diff
@@ -275,6 +296,9 @@ class Nokia_LCD {
      */
     bool printCharacter(char character);
 
+protected:
+    bool coupled = false;
+private:
```
Should be `private` and use the `m` prefix, i.e. `mCoupled`.
```suggestion
    bool mCoupled = false;
```

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -291,7 +304,9 @@ bool Nokia_LCD::send(const unsigned char lcd_byte, const bool is_data,
 
     // Send the byte
-    digitalWrite(kCe_pin, LOW);
+    if (!coupled) {
```
```suggestion
    if (!mCoupled) {
```

platisd commented on src/Nokia_LCD.cpp:
```diff
@@ -301,7 +316,9 @@ bool Nokia_LCD::send(const unsigned char lcd_byte, const bool is_data,
     }
 
-    digitalWrite(kCe_pin, HIGH);
+    if (!coupled) {
```
```suggestion
    if (!mCoupled) {
```

platisd reviewed (changes requested):
I think we are getting somewhere :)

jjtmp committed:
Apply suggestions from code review

platisds' points are reasonable; you can manage to use couple()/uncouple() only when needed; the API description can be a straightforward functional description with little example.

Co-authored-by: Dimitris Platis <platisd@users.noreply.github.com>

jjtmp commented:
It is right that the performance claim isn't from measurements and that the interface description is better brief than possibly wrong. One could argue how much access to inner variables should be possible, but of course it's not unacceptable to make access private.
I committed your suggestions.

platisd reviewed (approved):
:+1: :tada:

jjtmp commented:
👍 Thank you! Makes exchange of the .ino file easier between my build pipeline (Makefile) and the project initiators build pipeline (Arduino IDE) without any explanation.
As soon as I have access to the hardware again I will check about the electrical interference topic, just to be sure.

jjtmp commented:
As signalized I have checked the "electrical interference" issue. Maybe it isn't related to electrical interference, at least I couldn't reproduce this. It's probably just a question of reset pins. The begin() method does set reset pin to low (and high) and then writes makes necessary initialization bytes. So there are following variants to share pins of two LCDs:

- [hardware] either use distinct RST pins
- [software] or set RST pin to the same value as CE pin in instantiation of LCD object (for those later LCD for which you call begin() after you called it for the first LCD)
```
Nokia_LCD lcd10(13, 12, 11, 10, 9);
Nokia_LCD lcd7(13, 12, 11, 7, 7); // reset pin "7" = trick the constructor
lcd10.begin()
lcd7.begin()
```
- [software] or use couple(). the couple() must be called for all LCDs. It is not necessary for the LCD you call begin() on. Example: 
```
Nokia_LCD lcd10(13, 12, 11, 10, 9);
Nokia_LCD lcd7(13, 12, 11, 7, 9); // reset pin irrelevant, never touched
lcd7.couple(); // new since version NokiaLCD 2.7.0
lcd10.begin();
lcd7.uncouple();
```
So _now_ I'm done with that story.

platisd commented:
Thanks for the follow-up @jjtmp :100:
